libvast/src/transform_steps/delete.cpp

```diff
@@ -17,30 +17,58 @@
 
 namespace vast {
 
-delete_step::delete_step(const std::string& fieldname) : fieldname_(fieldname) {
+delete_step::delete_step(std::vector<std::string> fields)
+  : fields_(std::move(fields)) {
 }
 
-caf::expected<table_slice> delete_step::operator()(table_slice&& slice) const {
-  const auto& layout = slice.layout();
+caf::expected<std::pair<vast::type, std::vector<int>>>
+delete_step::adjust_layout(const vast::type& layout) const {
+  auto to_remove = std::set<vast::offset>{};
   const auto& layout_rt = caf::get<record_type>(layout);
-  auto offset = layout_rt.resolve_key(fieldname_);
-  if (!offset)
-    return std::move(slice);
-  auto columnn_index = layout_rt.flat_index(*offset);
-  auto adjusted_layout_rt
-    = layout_rt.transform({{*offset, record_type::drop()}});
+  for (const auto& key : fields_) {
+    auto offsets = layout_rt.resolve_key_suffix(key);
+    for (const auto& offset : offsets) {
+      to_remove.emplace(offset);
+    }
+  }
+  if (to_remove.empty())
+    return caf::no_error;
+  // adjust layout
+  auto transformations = std::vector<record_type::transformation>{};
+  auto flat_index_to_keep = std::vector<int>{};
+  for (const auto& [field, offset] : layout_rt.leaves()) {
+    if (to_remove.contains(offset))
+      transformations.push_back({offset, record_type::drop()});
+    else
+      flat_index_to_keep.push_back(
+        static_cast<int>(layout_rt.flat_index(offset)));
+  }
```

Comment:
```suggestion
  for (auto flat_index = 0;
       const auto& [field, offset] : layout_rt.leaves()) {
    if (to_remove.contains(offset))
      transformations.push_back({offset, record_type::drop()});
    else
      flat_index_to_keep.push_back(flat_index);
    ++flat_index;
  }
```

If you're iterating over all the leaves anyways, you might as well keep track of the running flat index. Then you don't need to convert the offset into a flat index at all.